src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -2194,12 +2195,23 @@ public override bool HandleCall(MethodIL callingMethodBody, MethodDesc calledMet
                                 }
 
                                 comDangerousMethod |= IsComInterop(marshalAsDescriptor, calledMethod.Signature[paramIndex]);
+                                abstractDelegatesMethod |= IsAbstractDelegatesInterop(marshalAsDescriptor, calledMethod.Signature[paramIndex]);
                             }
 
                             if (comDangerousMethod)
                             {
                                 reflectionContext.AnalyzingPattern();
-                                reflectionContext.RecordUnrecognizedPattern(2050, $"P/invoke method '{calledMethod.GetDisplayName()}' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed.");
+                                reflectionContext.RecordUnrecognizedPattern(
+                                    (int)DiagnosticId.CorrectnessOfCOMCannotBeGuaranteed, 
+                                    new DiagnosticString(DiagnosticId.CorrectnessOfCOMCannotBeGuaranteed).GetMessage(DiagnosticUtilities.GetMethodSignatureDisplayName(calledMethod)));
+                            }
+
+                            if (abstractDelegatesMethod)
+                            {
+                                reflectionContext.AnalyzingPattern();
+                                reflectionContext.RecordUnrecognizedPattern(
+                                    (int)DiagnosticId.CorrectnessOfAbstractDelegatesCannotBeGuaranteed, 
+                                    new DiagnosticString(DiagnosticId.CorrectnessOfAbstractDelegatesCannotBeGuaranteed).GetMessage(DiagnosticUtilities.GetMethodSignatureDisplayName(calledMethod)));
```

Comment:
Missed this earlier:

```suggestion
                                _logger.LogWarning(DiagnosticString(DiagnosticId.CorrectnessOfAbstractDelegatesCannotBeGuaranteed).GetMessage(DiagnosticUtilities.GetMethodSignatureDisplayName(calledMethod), (int)DiagnosticId.CorrectnessOfAbstractDelegatesCannotBeGuaranteed, callingMethodBody, offset, MessageSubCategory.AotAnalysis);
```

The `AnalyzingPattern` stuff is only for trimming analysis - this is not trimming analysis.

Reply:
Without `AnalyzingPattern` exception happens in DEBUG mode.

Reply:
I meant the whole `ReflectionContext`, not just `AnalyzingPattern` - this should call into logger directly instead of using `RecordUnrecognizedPattern` - the suggested code snippet above should be close enough.

`ReflectionContext` will end up logging too, but it reports trim analysis warnings, not AOT warnings (the category is wrong).